arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,476 @@
+#define L1_SPACE "  "
+#define L2_SPACE "    "
+#define L3_SPACE "      "
+#define XLAT_TABLE_LEVEL_SPACE(level)		\
+	(((level) == 0) ? L0_SPACE :		\
+	((level) == 1) ? L1_SPACE :		\
+	((level) == 2) ? L2_SPACE : L3_SPACE)
+#endif
+
+static u64_t base_xlat_table[NUM_BASE_LEVEL_ENTRIES]
+		__aligned(NUM_BASE_LEVEL_ENTRIES * sizeof(u64_t));
+
+static u64_t xlat_tables[CONFIG_MAX_XLAT_TABLES][XLAT_TABLE_ENTRIES]
+		__aligned(XLAT_TABLE_ENTRIES * sizeof(u64_t));
+
+/* Translation table control register settings */
+static u64_t get_tcr(int el)
+{
+	u64_t tcr;
+	u64_t pa_bits = CONFIG_ARM64_PA_BITS;
+	u64_t va_bits = CONFIG_ARM64_VA_BITS;
+	u64_t tcr_ps_bits;
+
+	switch (pa_bits) {
+	case 48:
+		tcr_ps_bits = TCR_PS_BITS_256TB;
+		break;
+	case 44:
+		tcr_ps_bits = TCR_PS_BITS_16TB;
+		break;
+	case 42:
+		tcr_ps_bits = TCR_PS_BITS_4TB;
+		break;
+	case 40:
+		tcr_ps_bits = TCR_PS_BITS_1TB;
+		break;
+	case 36:
+		tcr_ps_bits = TCR_PS_BITS_64GB;
+		break;
+	default:
+		tcr_ps_bits = TCR_PS_BITS_4GB;
+		break;
+	}
+
+	if (el == 1) {
+		tcr = (tcr_ps_bits << TCR_EL1_IPS_SHIFT);
+		/*
+		 * TCR_EL1.EPD1: Disable translation table walk for addresses
+		 * that are translated using TTBR1_EL1.
+		 */
+		tcr |= TCR_EPD1_DISABLE;
+	} else
+		tcr = (tcr_ps_bits << TCR_EL3_PS_SHIFT);
+
+	tcr |= TCR_T0SZ(va_bits);
+	/*
+	 * Translation table walk is cacheable, inner/outer WBWA and
+	 * inner shareable
+	 */
+	tcr |= TCR_TG0_4K | TCR_SHARED_INNER | TCR_ORGN_WBWA | TCR_IRGN_WBWA;
+
+	return tcr;
+}
+
+static int pte_desc_type(u64_t *pte)
+{
+	return *pte & PTE_DESC_TYPE_MASK;
+}
+
+static u64_t *calculate_pte_index(u64_t addr, int level)
+{
+	int base_level = XLAT_TABLE_BASE_LEVEL;
+	u64_t *pte;
+	u64_t idx;
+	unsigned int i;
+
+	/* Walk through all translation tables to find pte index */
+	pte = (u64_t *)base_xlat_table;
+	for (i = base_level; i <= XLAT_TABLE_LEVEL_MAX; i++) {
+		idx = XLAT_TABLE_VA_IDX(addr, i);
+		pte += idx;
+
+		/* Found pte index */
+		if (i == level)
+			return pte;
+		/* if PTE is not table desc, can't traverse */
+		if (pte_desc_type(pte) != PTE_TABLE_DESC)
+			return NULL;
+		/* Move to the next translation table level */
+		pte = (u64_t *)(*pte & 0x0000fffffffff000ULL);
+	}
+
+	return NULL;
+}
+
+static void set_pte_table_desc(u64_t *pte, u64_t *table, unsigned int level)
+{
+#if DUMP_PTE
+	MMU_DEBUG("%s", XLAT_TABLE_LEVEL_SPACE(level));
+	MMU_DEBUG("%p: [Table] %p\n", pte, table);
+#endif
+	/* Point pte to new table */
+	*pte = PTE_TABLE_DESC | (u64_t)table;
+}
+
+static void set_pte_block_desc(u64_t *pte, u64_t addr_pa,
+			       unsigned int attrs, unsigned int level)
+{
+	u64_t desc = addr_pa;
+	unsigned int mem_type;
+
+	desc |= (level == 3) ? PTE_PAGE_DESC : PTE_BLOCK_DESC;
+
+	/* NS bit for security memory access from secure state */
+	desc |= (attrs & MT_NS) ? PTE_BLOCK_DESC_NS : 0;
+
+	/* AP bits for Data access permission */
+	desc |= (attrs & MT_RW) ? PTE_BLOCK_DESC_AP_RW : PTE_BLOCK_DESC_AP_RO;
+
+	/* the access flag */
```

Comment:
The “The AP[2:1] data access permissions bits” must enable using MMU from EL0 also. 

#define PTE_BLOCK_DESC_EL0_EN                          (1UL << 6)
/* AP bits for EL0/El1 access permission */
desc |= PTE_BLOCK_DESC_EL0_EN;

Reply:
Normally, we would be executing user applications at EL0. Currently, I do not see any EL0 software getting executed with zephyr. So, I am not sure if this change should be done.

Reply:
@carlocaione What the execution level for user task in current aarch64 implementation?

Reply:
Everything is run at EL3